File: src/chunking/token_window_chunker.py
from src.chunking.base import Chunker
from src.models import DocumentPage, Section
from src.tokenizer import TiktokenCounter
DEFAULT_MAX_TOKENS = 400
DEFAULT_OVERLAP_TOKENS = 60
class TokenWindowChunker(Chunker):
    def __init__(
        self,
        max_tokens: int = DEFAULT_MAX_TOKENS,
        overlap_tokens: int = DEFAULT_OVERLAP_TOKENS,
        counter: TiktokenCounter | None = None,
    ):
        self.max_tokens = max_tokens
        self.overlap_tokens = overlap_tokens
        self.counter = counter or TiktokenCounter()

    def windows(self, text: str) -> list[str]:
        blocks = [b.strip() for b in text.split("\n") if b.strip()]
        if not blocks:
            return []
        sized = [(b, self.counter.count(b)) for b in blocks]
        windows: list[str] = []
        current: list[tuple[str, int]] = []
        total = 0
        for block, size in sized:
            if current and total + size > self.max_tokens:
                windows.append("\n".join(b for b, _ in current))
                carried: list[tuple[str, int]] = []
                carried_total = 0
                for item in reversed(current):
                    if carried_total + item[1] > self.overlap_tokens:
                        break
                    carried.insert(0, item)
                    carried_total += item[1]
                current, total = carried, carried_total
            current.append((block, size))
            total += size
        if current:
            windows.append("\n".join(b for b, _ in current))
        return windows

    def split(self, pages: list[DocumentPage]) -> list[Section]:
        if not pages:
            return []
        source = pages[0].source
        text = "\n".join(page.text for page in pages)
        sections = []
        for window in self.windows(text):
            sections.append(
                Section(
                    section_id = f"s{len(sections) + 1:02d}",
                    title = f"{source} (part {len(sections) + 1})",
                    text = window,
                    source = source,
                    page_number = pages[0].page_number,
                    parent_heading = None,
                    token_count = self.counter.count(window),
                )
            )
        return sections

File: src/chunking/tokenWindowChunker.ts
import { TiktokenCounter } from "../tokenizer.ts";
import type { Chunker } from "./base.ts";
import type { DocumentPage, Section } from "../models.ts";
import type { TokenCounter } from "../tokenizer.ts";

const DEFAULT_MAX_TOKENS = 400;
const DEFAULT_OVERLAP_TOKENS = 60;

export interface TokenWindowOptions {
  maxTokens?: number;
  overlapTokens?: number;
  counter?: TokenCounter;
}

/** The fallback splitter: fixed-size windows with a little overlap, split on blank
 *  lines so a window never starts mid-sentence. */
export class TokenWindowChunker implements Chunker {
  readonly maxTokens: number;
  readonly overlapTokens: number;
  readonly counter: TokenCounter;

  constructor(options: TokenWindowOptions = {}) {
    this.maxTokens = options.maxTokens ?? DEFAULT_MAX_TOKENS;
    this.overlapTokens = options.overlapTokens ?? DEFAULT_OVERLAP_TOKENS;
    this.counter = options.counter ?? new TiktokenCounter();
  }

  windows(text: string): string[] {
    const blocks = text
      .split("\n")
      .map((block) => block.trim())
      .filter(Boolean);
    if (blocks.length === 0) return [];

    const sized = blocks.map((block) => ({ block, size: this.counter.count(block) }));
    const windows: string[] = [];
    let current: { block: string; size: number }[] = [];
    let total = 0;

    for (const item of sized) {
      if (current.length > 0 && total + item.size > this.maxTokens) {
        windows.push(current.map((c) => c.block).join("\n"));
        // carry the tail of this window into the next one, so a sentence that
        // straddles the boundary is still findable from either side
        const carried: { block: string; size: number }[] = [];
        let carriedTotal = 0;
        for (const previous of [...current].reverse()) {
          if (carriedTotal + previous.size > this.overlapTokens) break;
          carried.unshift(previous);
          carriedTotal += previous.size;
        }
        current = carried;
        total = carriedTotal;
      }
      current.push(item);
      total += item.size;
    }
    if (current.length > 0) windows.push(current.map((c) => c.block).join("\n"));
    return windows;
  }

  split(pages: DocumentPage[]): Section[] {
    const first = pages[0];
    if (!first) return [];
    const text = pages.map((page) => page.text).join("\n");

    return this.windows(text).map((window, index) => ({
      sectionId: `s${String(index + 1).padStart(2, "0")}`,
      title: `${first.source} (part ${index + 1})`,
      text: window,
      source: first.source,
      pageNumber: first.pageNumber,
      parentHeading: null,
      tokenCount: this.counter.count(window),
    }));
  }
}
